native/service/cross_chain_manager/ripple/ripple_handler.go

```diff
@@ -0,0 +1,285 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The  poly network  is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with The poly network .  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package ripple
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"math/big"
+
+	"github.com/polynetwork/poly/common"
+	"github.com/polynetwork/poly/native"
+	"github.com/polynetwork/poly/native/event"
+	crosscommon "github.com/polynetwork/poly/native/service/cross_chain_manager/common"
+	"github.com/polynetwork/poly/native/service/governance/side_chain_manager"
+	"github.com/polynetwork/poly/native/service/utils"
+	"github.com/polynetwork/ripple-sdk/types"
+	"github.com/rubblelabs/ripple/data"
+)
+
+type RippleHandler struct {
+}
+
+func NewRippleHandler() *RippleHandler {
+	return &RippleHandler{}
+}
+
+func (this *RippleHandler) MultiSign(service *native.NativeService) error {
+	params := new(MultiSignParam)
+	if err := params.Deserialization(common.NewZeroCopySource(service.GetInput())); err != nil {
+		return fmt.Errorf("MultiSign, contract params deserialize error: %v", err)
+	}
+
+	// get rippleExtraInfo
+	rippleExtraInfo, err := side_chain_manager.GetRippleExtraInfo(service, params.ToChainId, params.AssetAddress)
+	if err != nil {
+		return fmt.Errorf("MultiSign, side_chain_manager.GetRippleExtraInfo error")
+	}
+
+	// get raw txJsonInfo
+	raw, err := GetTxJsonInfo(service, params.FromChainId, params.TxHash)
+	if err != nil {
+		return fmt.Errorf("MultiSign, get txJsonInfo error")
+	}
+
+	// check if aleady done
+	multisignInfo, err := GetMultisignInfo(service, raw)
+	if err != nil {
+		return fmt.Errorf("MultiSign, GetMultisignInfo error")
+	}
+	if multisignInfo.Status {
+		return nil
+	}
+
+	// check if signature is valid
+	txJson := new(types.MultisignPayment)
+	err = json.Unmarshal([]byte(params.TxJson), txJson)
+	if err != nil {
+		return fmt.Errorf("MultiSign, unmarshal signed txjson error: %s", err)
+	}
+	for _, s := range txJson.Signers {
+		signerAccount, err := data.NewAccountFromAddress(s.Signer.Account)
+		if err != nil {
+			return fmt.Errorf("MultiSign, data.NewAccountFromAddress error: %s", err)
+		}
+		signerPk, err := hex.DecodeString(s.Signer.SigningPubKey)
+		if err != nil {
+			return fmt.Errorf("MultiSign, hex.DecodeString signer pk error: %s", err)
+		}
+		signature, err := hex.DecodeString(s.Signer.TxnSignature)
+		if err != nil {
+			return fmt.Errorf("MultiSign, hex.DecodeString signature error: %s", err)
+		}
+
+		// check if valid signer
+		flag := false
+		for _, v := range rippleExtraInfo.Pks {
+			if fmt.Sprintf("%X", v) == s.Signer.SigningPubKey {
+				flag = true
```

Comment:
break missing? convert s.Signer.SigningPubKey to same format?

Reply:
Here check s.Signer.SigningPubKey is the specific multi signer's publickey, so convert s.Signer.SigningPubKey to same format and check if s.Signer.SigningPubKey is in multi signer group.